### DISCUSSION TITLE
Improve isEmpty types

### DISCUSSION THREAD

**Discussion by MarlonPassos-git** [radashi-org/radashi#209] [Ideas]
## Sumary

Looking for the [lodash types of `isEmpty`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/common/lang.d.ts#L574) they are well advanced.

We could work on the `value is` to improve typing.

it would be great if typing reached this level:

```ts
type User = {
  name: string
  age: number
}

const list: User[] = []
if (_.isEmpty(list)) {
  list // list is undefined[]
} else {
  list // list is User[] 
}

const str = ''
if (_.isEmpty(str)) {
  str // str is ''  
} else {
  str // str is string
}

const obj = {} as User

if (_.isEmpty(obj)) {
  obj // obj is {}
} else {
  obj // obj is User
}
```

## related Issue
https://github.com/sodiray/radash/issues/422

**Comment by aleclarson** [comment]
Type guards can only narrow the type, so `User` to `{}` isn't possible (`User` is a stricter type). Btw, using the `{}` type is bad practice, because even primitives can be assigned to it (everything is an object in JavaScript after all!). The TypeScript team recommends using `object` instead.

For arrays, the more accurate type is `never[]`. If `undefined[]` was used, the narrowed type in your example would be `User[] & undefined[]`, so `list[0]` would result in `never` anyway.

**Reply by MarlonPassos-git** [reply]
You are right. We can use these types that you mentioned.

**Comment by aleclarson** [comment]
Addressed in #219
